Index: class/System/System.ComponentModel/ListChangedEventArgs.cs
===================================================================
--- class/System/System.ComponentModel/ListChangedEventArgs.cs	(revision 114541)
+++ class/System/System.ComponentModel/ListChangedEventArgs.cs	(working copy)
@@ -45,6 +45,7 @@
		{
			this.changedType = listChangedType;
			this.newIndex = newIndex;
+			this.oldIndex = -1;
		}

		public ListChangedEventArgs (ListChangedType listChangedType,
