Index: class/corlib/System/DateTime.cs
===================================================================
--- class/corlib/System/DateTime.cs	(revision 114651)
+++ class/corlib/System/DateTime.cs	(working copy)
@@ -1523,7 +1523,7 @@
					else if (num == 1)
						tzoffset = _ParseNumber (s, valuePos, 1, 2, true, sloppy_parsing, out num_parsed);
					else {
-						tzoffset = _ParseNumber (s, valuePos, 1, 2, true, sloppy_parsing, out num_parsed);
+						tzoffset = _ParseNumber (s, valuePos, 1, 2, true, /*sloppy_parsing*/true, out num_parsed);
						valuePos += num_parsed;
						if (num_parsed < 0)
							return false;
