Index: class/corlib/System.Collections.Generic/Dictionary.cs
===================================================================
--- class/corlib/System.Collections.Generic/Dictionary.cs	(revision 114541)
+++ class/corlib/System.Collections.Generic/Dictionary.cs	(working copy)
@@ -398,6 +398,9 @@

		public bool ContainsKey (TKey key)
		{
+			if (key == null)
+				throw new ArgumentNullException ("key");
+
			// get first item of linked list corresponding to given key
			int hashCode = hcp.GetHashCode (key);
			int cur = table [(hashCode & int.MaxValue) % table.Length] - 1;
